pallets/qa-tools/src/lib.rs

```diff
@@ -0,0 +1,394 @@
+                (1, balance!(85.4)),
+                (1, balance!(93.2)),
+                (3, balance!(36.6)),
+                (3, balance!(205.5)),
+                (3, balance!(13.7)),
+            ];
+
+            let mut data = order_book::cache_data_layer::CacheDataLayer::<T>::new();
+
+            for (order_book_id, settings) in fill_settings {
+                Self::fill_order_book(
+                    &mut data,
+                    order_book_id,
+                    asks_owner.clone(),
+                    bids_owner.clone(),
+                    buy_orders_steps.into_iter(),
+                    sell_orders_steps.into_iter(),
+                    settings,
+                    now,
+                )?;
+            }
+            data.commit();
+            Ok(())
+        }
+
+        /// Fill a single order book.
+        fn fill_order_book(
+            data: &mut CacheDataLayer<T>,
+            book_id: OrderBookId<T::AssetId>,
+            asks_owner: T::AccountId,
+            bids_owner: T::AccountId,
+            buy_orders_steps: impl Iterator<Item = (u128, Balance)>,
+            sell_orders_steps: impl Iterator<Item = (u128, Balance)>,
+            settings: OrderBookFillSettings,
+            now: MomentOf<T>,
+        ) -> Result<(), DispatchError> {
+            let current_block = frame_system::Pallet::<T>::block_number();
+            let mut order_book = <order_book::OrderBooks<T>>::get(book_id)
+                .ok_or(Error::<T>::OrderBookUnkonwnBook)?;
+
+            // Convert price steps and best price to actual prices
+            let buy_orders: Vec<_> = buy_orders_steps
+                .map(|(price_steps, base)| {
+                    (
+                        settings.best_bid_price - price_steps * order_book.tick_size,
+                        order_book.align_amount(base + order_book.step_lot_size),
+                    )
+                })
+                .collect();
+            let sell_orders: Vec<_> = sell_orders_steps
+                .map(|(price_steps, base)| {
+                    (
+                        settings.best_ask_price + price_steps * order_book.tick_size,
+                        order_book.align_amount(base + order_book.step_lot_size),
+                    )
+                })
+                .collect();
+            // Total amount of quote asset to be locked from `bids_owner`
+            let buy_quote_locked: Balance = buy_orders
+                .iter()
+                .map(|(quote, base)| {
+                    let quote_amount_fixed = FixedWrapper::from(*quote) * FixedWrapper::from(*base);
+                    quote_amount_fixed.into_balance()
+                })
+                .sum();
+            let sell_base_locked: Balance = sell_orders.iter().map(|(_, base)| base).sum();
+
+            // mint required amount to make this extrinsic self-sufficient
+            assets::Pallet::<T>::mint_unchecked(&book_id.quote, &bids_owner, buy_quote_locked)?;
```

Comment:
Usually we use update_balance

Reply:
`update_own_balance` requires mint permissions for the account, the extrinsic (`update_balance`) is undesirable to call, and I recall we wanted to use `assets` pallet as a single interface to operate with the balances

I'm not sure how to do it otherwise

Reply:
ok